catalogs/catalog-lakehouse-paimon/src/main/java/org/apache/gravitino/catalog/lakehouse/paimon/utils/PaimonHiveCachedClientPool.java

```diff
@@ -0,0 +1,304 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ *
+ */
+package org.apache.gravitino.catalog.lakehouse.paimon.utils;
+
+import static org.apache.paimon.hive.HiveCatalogOptions.CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS;
+import static org.apache.paimon.hive.HiveCatalogOptions.CLIENT_POOL_CACHE_KEYS;
+import static org.apache.paimon.options.CatalogOptions.CLIENT_POOL_SIZE;
+
+import java.io.IOException;
+import java.io.UncheckedIOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Locale;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Set;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import javax.annotation.Nullable;
+import org.apache.hadoop.conf.Configuration;
+import org.apache.hadoop.hive.conf.HiveConf;
+import org.apache.hadoop.hive.metastore.IMetaStoreClient;
+import org.apache.hadoop.security.UserGroupInformation;
+import org.apache.paimon.annotation.VisibleForTesting;
+import org.apache.paimon.client.ClientPool;
+import org.apache.paimon.hive.HiveCatalogOptions;
+import org.apache.paimon.hive.pool.HiveClientPool;
+import org.apache.paimon.options.Options;
+import org.apache.paimon.shade.caffeine2.com.github.benmanes.caffeine.cache.Cache;
+import org.apache.paimon.shade.caffeine2.com.github.benmanes.caffeine.cache.Caffeine;
+import org.apache.paimon.shade.caffeine2.com.github.benmanes.caffeine.cache.Scheduler;
+import org.apache.paimon.shade.guava30.com.google.common.collect.Lists;
+import org.apache.paimon.shade.guava30.com.google.common.collect.Maps;
+import org.apache.paimon.shade.guava30.com.google.common.collect.Sets;
+import org.apache.paimon.shade.guava30.com.google.common.util.concurrent.ThreadFactoryBuilder;
+import org.apache.paimon.utils.Preconditions;
+import org.apache.thrift.TException;
+
+/**
+ * Referred from Apache Paimon's CachedClientPool implementation
+ * paimon-hive/paimon-hive-catalog/src/main/java/org/apache/paimon/hive/pool/CachedClientPool.java
+ */
+public class PaimonHiveCachedClientPool implements ClientPool<IMetaStoreClient, TException> {
+
+  private static final String CONF_ELEMENT_PREFIX = "conf:";
+
+  private static Cache<Key, HiveClientPool> clientPoolCache;
+
+  private final Configuration conf;
+  private final int clientPoolSize;
+  private final long evictionInterval;
+  private final Key key;
+  private final String clientClassName;
+
+  public PaimonHiveCachedClientPool(Configuration conf, Options options, String clientClassName) {
+    this.conf = conf;
+    this.clientPoolSize = options.get(CLIENT_POOL_SIZE);
+    this.evictionInterval = options.get(CLIENT_POOL_CACHE_EVICTION_INTERVAL_MS);
+    this.key = extractKey(clientClassName, options.get(CLIENT_POOL_CACHE_KEYS), conf);
```

Comment:
So the key is a constant value, consider the following scenario:

User1 access the Paimon catalog then a connection pool with the key was created, when user2 also wants to access the catalog, it will reuse the pool created by user1. Since connection contains user information, it will cause issues.